core/store/src/flat_state.rs

```diff
@@ -60,20 +97,79 @@ mod imp {
         }
     }
 
-    /// Possibly creates a new [`FlateState`] object backed by given storage.
-    ///
-    /// Always returns `None` if the `protocol_feature_flat_state` Cargo feature is
-    /// not enabled.  Otherwise, returns a new [`FlatState`] object backed by
-    /// specified storage if `use_flat_state` argument is true.
-    pub fn maybe_new(use_flat_state: bool, store: &Store) -> Option<FlatState> {
-        use_flat_state.then(|| FlatState { store: store.clone() })
+    /// `FlatStateFactory` provides a way to construct new flat state to pass to new tries.
+    /// It is owned by NightshadeRuntime, and thus can be owned by multiple threads, so the implementation
+    /// must be thread safe.
+    #[derive(Clone)]
+    pub struct FlatStateFactory(Arc<FlatStateFactoryInner>);
+
+    pub struct FlatStateFactoryInner {
+        store: Store,
+        caches: Mutex<HashMap<ShardId, FlatStateCache>>,
+        /// Here we store the flat_storage_state per shard. The reason why we don't use the same
+        /// FlatStorageState for all shards is that there are two modes of block processing,
+        /// normal block processing and block catchups. Since these are performed on different range
+        /// of blocks, we need flat storage to be able to support different range of blocks
+        /// on different shards. So we simply store a different state for each shard.
+        /// This may cause some overhead because the data like shards that the node is processing for
+        /// this epoch can share the same `head` and `tail`, similar for shards for the next epoch,
+        /// but such overhead is negligible comparing the delta sizes, so we think it's ok.
+        flat_storage_states: Mutex<HashMap<ShardId, FlatStorageState>>,
+    }
+
+    impl FlatStateFactory {
+        pub fn new(store: Store) -> Self {
+            Self(Arc::new(FlatStateFactoryInner {
+                store,
+                caches: Default::default(),
+                flat_storage_states: Default::default(),
+            }))
+        }
+
+        pub fn new_flat_state_for_shard(
+            &self,
+            shard_id: ShardId,
+            is_view: bool,
+        ) -> Option<FlatState> {
+            if is_view {
+                // TODO: Technically, like TrieCache, we should have a separate set of caches for Client and
+                // ViewClient. Right now, we can get by by not enabling flat state for view trie
+                None
+            } else {
+                let cache = {
+                    let mut caches = self.0.caches.lock().expect(POISONED_LOCK_ERR);
```

Comment:
It's better to use just `.unwrap()` for poisoning errors. We compile with `panic=abort`, so it's impossible to even get a poisoning error (as the thread which panicked would abort the whole process)

Reply:
Will change it in a following PR